### PR TITLE
[SPARK-30879][DOCS] Refine workflow for building docs

### PR DESCRIPTION
**Pull request by nchammas**
### What changes were proposed in this pull request?

This PR makes the following refinements to the workflow for building docs:
* Install Python and Ruby consistently using pyenv and rbenv across both the docs README and the release Dockerfile.
* Pin the Python and Ruby versions we use.
* Pin all direct Python and Ruby dependency versions.
* Eliminate any use of `sudo pip`, which the Python community discourages, or `sudo gem`.

### Why are the changes needed?

This PR should increase the consistency and reproducibility of the doc-building process by managing Python and Ruby in a more consistent way, and by eliminating unused or outdated code.

Here's a possible example of an issue building the docs that would be addressed by the changes in this PR: https://github.com/apache/spark/pull/27459#discussion_r376135719

### Does this PR introduce any user-facing change?

No.

### How was this patch tested?

Manual tests:
* I was able to build the Docker image successfully, minus the final part about `RUN useradd`.
    * I am unable to run `do-release-docker.sh` because I am not a committer and don't have the required GPG key.
* I built the docs locally and viewed them in the browser.

I think I need a committer to more fully test out these changes.

### CODE REVIEW

File: sql/mkdocs.yml
Comment (HyukjinKwon) [inline sql/mkdocs.yml]: Yeap, I saw the warning. +1

File: python/docs/Makefile
Comment (HyukjinKwon) [inline python/docs/Makefile]: +1 for this change too although I didn't check the details. This Makefile was from the old sphinx example.

File: dev/create-release/spark-rm/Dockerfile
Comment (HyukjinKwon) [inline dev/create-release/spark-rm/Dockerfile]: @dongjoon-hyun and @wangyum might better to review.

File: docs/README.md
Comment (HyukjinKwon) [inline docs/README.md]: @nchammas, btw why should we fix the version? I think it's better to promote the latest versions.

File: docs/README.md
Comment (HyukjinKwon) [inline docs/README.md]: @nchammas, what about we keep the default way, and simply mention such pyend, virtualenv or conda?

File: docs/README.md
Comment (nchammas) [reply docs/README.md]: We should pin dependencies for the usual reason: It guarantees a stable environment that gives reproducible results.

New releases may have breaking behavior, so you don't always want to install the latest version. And different users building the docs may have outdated versions installed, which will yield them a broken build. (e.g. If you ran `pip install sphinx` when the latest release was 1.1.0, then pip will not update your version of sphinx as new releases are made unless you explicitly request it via `pip install --upgrade`.)

These categories of build problem are addressed by pinning versions.

File: docs/README.md
Comment (nchammas) [reply docs/README.md]: We can do that, but I am promoting pyenv and rbenv because they give us a consistent way to install any version of Python and Ruby across Docker and the typical developer environment, without depending on package repositories that typically lag a year or two behind the latest release. pyenv and rbenv also give and simpler and more consistent workflow for getting and using specific versions of Python and Ruby.

With the way things are right now, we don't specify any version of Python or Ruby in the docs README, and in the Dockerfile we use a [4 year old version of Ruby](https://www.ruby-lang.org/en/news/2015/12/25/ruby-2-3-0-released/) and [3 year old version of Python](https://www.python.org/downloads/release/python-360/). Furthermore, to upgrade from these versions to newer ones, developers have to figure out the upgrade path themselves for their local workstation, and for Docker they have to hunt through package repositories that, again, often lag behind what's available.

pyenv and rbenv give us a consistent way to install and manage what versions of Ruby and Python we're using to build docs -- whether on the local machine or in a Docker image -- and they make it dead simple to change the versions we want to use.

**Comment by nchammas** [comment]
Should I break up this PR to make it easier to review? e.g.
1. Makefile trim
2. pages -> nav
3. pyenv and rbenv + dependency pinning

Asking just in case the silence from other reviewers is because this PR is too big.

If y'all just haven't gotten around to taking a look, no worries. I'm being proactive.

**Comment by nchammas** [comment]
I've split off the Makefile and Mkdocs changes and updated my previous comment with links to the new PRs.

**Comment by nchammas** [comment]
Retest this please.

**Comment by nchammas** [comment]
Is it possible for contributors to re-trigger the GitHub workflows for a certain commit? I'm not seeing how to do that.

File: docs/README.md
Comment (nchammas) [inline docs/README.md]: One alternative to running these `init` commands is to prefix the below commands so that they do the right thing without depending on any environment variables being set by `init`.

So, for example, `pip install` would become `pyenv exec pip install` and `jekyll` would become `rbenv exec jekyll`.

**Comment by dongjoon-hyun** [comment]
cc @vanzin for `spark-rm` changes.

**Review by dongjoon-hyun — CHANGES_REQUESTED** [review]
Please don't reuse the same JIRA ID for different PR, @nchammas .
- https://github.com/apache/spark/pull/27626 [SPARK-30731] Update deprecated Mkdocs option
- https://github.com/apache/spark/pull/27625 [SPARK-30731] Delete Sphinx Makefile cruft
- https://github.com/apache/spark/pull/27534 [SPARK-30731][DOCS] Refine workflow for building docs

**Comment by dongjoon-hyun** [comment]
Since https://github.com/apache/spark/pull/27626 is merged to `branch-2.4` as SPARK-30731, you had better update SPARK-30731 accordingly. And, this PR should have different new JIRA ID.

**Comment by nchammas** [comment]
@dongjoon-hyun 

> Please don't reuse the same JIRA ID for different PR

The 3 PRs were initially 1 PR. I split up this PR into 3 to make it easier to review. The JIRA ticket captures what all 3 PRs are addressing, which is basically a set of refinements to the doc-building workflow.

As far as I know, there doesn't have to be a 1-to-1 relationship between PRs and JIRA tickets. Is that not the case?

**Comment by dongjoon-hyun** [comment]
@nchammas . No, you had better make it umbrella JIRA.
> The 3 PRs were initially 1 PR. I split up this PR into 3 to make it easier to review. The JIRA ticket captures what all 3 PRs are addressing, which is basically a set of refinements to the doc-building workflow.

First of all, after merging your three PRs, do you think what is going to be the `Fixed Version` of SPARK-30731? Do you think we should have `Fixed Version` for `2.4.6` because of https://github.com/apache/spark/commit/6e0c1162bf95938081e93c77a39b2826fec9c84e ?

Second, you created SPARK-30731 as `Improvement` JIRA. Did you aim to land it `branch-2.4` from the beginning? Then, you must understand Apache Spark community policy. We don't backport the improvement.

Since you are reusing the same JIRA ID, @HyukjinKwon mistakenly landed SPARK-30731 to `branch-2.4`.
- https://github.com/apache/spark/commit/6e0c1162bf95938081e93c77a39b2826fec9c84e

It seems that you are intentional because you don't understand the community policy, let me fix the situation first.

**Comment by dongjoon-hyun** [comment]
Hi, @HyukjinKwon .
We had better avoid landing this improvement on `branch-3.0`. This changes the release stuff a lot and and technically give a big risk to the next release manager because this resets our previous efforts and validation during preview/preview2.

**Comment by nchammas** [comment]
> Second, you created SPARK-30731 as Improvement JIRA. Did you aim to land it branch-2.4 from the beginning? Then, you must understand Apache Spark community policy. We don't backport the improvement.

No, I didn't aim to land it in branch-2.4. As I noted over on https://github.com/apache/spark/pull/27626#issuecomment-588391223, I don't know why the MkDocs patch was landed there. I create the JIRA with an "affects version" of 3.0.0 and didn't mention 2.4 anywhere. I didn't ask to backport any improvements, and I don't know why branch-2.4 became a point of discussion in any of these 3 PRs.

> It seems that you are intentional because you don't understand the community policy, let me fix the situation first.

I'm not sure what you mean by this, but I think you've imputed some motives to me that I don't have. All I wanted to do was split up 1 PR into 3 PRs. They are all refinements to the doc-building workflow. They are all improvements that don't need to go anywhere other than to master and maybe also branch-3.0.

I wasn't trying to circumvent community policy in any way. I'm happy to create any new JIRAs required, but it looks like you've gone ahead and done that yourself.

**Comment by nchammas** [comment]
> Do you think we should have Fixed Version for 2.4.6 because of 6e0c116 ?

I don't know why you are asking me this question. If you are implying that I somehow asked for there to be a new 2.4.x release because of 6e0c116, then there is some gross misunderstanding here. I have no opinions on the releases or where each patch should land, and I did not mention 2.4 in my JIRA ticket or any of the 3 PRs.

**Comment by dongjoon-hyun** [comment]
Got it. Then, it might be some misunderstanding. I thought you are aware of the community policies because you have been working in the community for a long time.
1. First of all, we don't backport `Improvement`. The decision is made per PR.
2. Due to (1), for improvement JIRAs, we had better use the version of master branch for `Affected Version`. In your case, you had better set `3.1.0` clearly when you register as `Improvement`.
3. For the same ID PRs, we use the same JIRA ID for the related follow-up PRs only. And, in that case, we use `[FOLLOWUP]` tag in the PR title.

Do you think your 3 PRs are follow-ups? For me, those are orthogonal. e.g., `Redefine workflow` is irrelevant to replace the deprecate `Mkdocs` option. The main reason why the committers ask to split is that you put orthogonal things into a single PR. It's bad not only for review, but also for understanding the commit logs.

**Comment by dongjoon-hyun** [comment]
I asked that questions because you insisted to have the same JIRA ID. You cannot share the same JIRA ID for three different your contributions because we will review per PRs.
> I don't know why you are asking me this question.

**Comment by nchammas** [comment]
> I thought you are aware of the community policies because you have been working in the community for a long time.

I thought I was familiar with the policies, too. 😄

> 1. First of all, we don't backport `Improvement`. The decision is made per PR.

Understood. And again, just to be clear, I didn't knowingly ask to backport any improvement.

> 2. Due to (1), for improvement JIRAs, we had better use the version of master branch for `Affected Version`. In your case, you had better set `3.1.0` clearly when you register as `Improvement`.

Perhaps this is part of why I am confused. By setting the "affects version" to 3.0.0 on my ticket, was I technically asking for the improvement to be backported to branch-3.0?

This still doesn't explain how branch-2.4 entered the picture, by the way. Is a backport to branch-2.4 implied because 3.0 hasn't been released yet?

> 3. For the same ID PRs, we use the same JIRA ID for the related follow-up PRs only. And, in that case, we use `[FOLLOWUP]` tag in the PR title.

I wasn't familiar with this convention, but: Duly noted.

> Do you think your 3 PRs are follow-ups? For me, those are orthogonal. e.g., `Redefine workflow` is irrelevant to replace the deprecate `Mkdocs` option. The main reason why the committers ask to split is that you put orthogonal things into a single PR. It's bad not only for review, but also for understanding the commit logs.

I agree. But just for the record, nobody asked me to split up the PR! I split it up on my own initiative to facilitate review. And in the process of doing that, as we've just seen, I unintentionally made a mess of things... 😅

File: docs/README.md
Comment (nchammas) [reply docs/README.md]: If the committers are not comfortable with this change, a more conservative change we can make for now is simply to pin the versions of the various Python and Ruby libraries that we install. We can revisit the rbenv/pyenv idea at a later time.

**Comment by dongjoon-hyun** [comment]
Oh, in that case, please use different IDs from the beginning next time especially.
> nobody asked me to split up the PR! 
Thanks, @nchammas .

**Comment by nchammas** [comment]
Will do. Apologies to @dongjoon-hyun and everyone else watching this PR for the noise.

File: docs/README.md
Comment (srowen) [reply docs/README.md]: I don't know much about these tools, but updating the docs to point to anything newer, if it sounds good to someone looking closely at this part, seems OK.

File: docs/README.md
Comment (srowen) [reply docs/README.md]: I think it's OK to pin. We have to maintain this across docs vs script, but, probably better this way.

**Review by srowen — COMMENTED** [review]
I don't quite follow the last 10 messages or so. @nchammas didn't back-port it, @HyukjinKwon did, but, it seems OK to do so IMHO: https://github.com/apache/spark/pull/27626 These things are all matters of judgment at some level.

Yes, we can land this only in master not 3.0, and I didn't see others suggest otherwise.

The various changes for this JIRA could be one PR, maybe better that way. Or individual JIRAs as they are only moderately related. But, for relatively minor ancillary changes, I don't think it matters all that much. The discussion is good to have but probably no big deal to proceed as-is now and resolve the JIRA once all are merged.

**Comment by HeartSaVioR** [comment]
Personally I'm feeling that setting Affected Version on improvement type is giving confusion while the value we get is relatively small, but let's discuss it in dev@ instead of here, as it's off-topic.

**Comment by nchammas** [comment]
I'd like to move this PR forward. What path would y'all like to take here?

Options are:
1. Keep discussing the possibility of using rbenv and pyenv, and continue to refine the PR as it is.
2. Scale back the PR to just pin the various Python and Ruby requirements to specific versions.
3. Close this PR and leave the doc building process as-is.

**Comment by srowen** [comment]
The change looks reasonable. Updating docs and pinning versions sounds good. It sounds like you were able to test most of it, and probably the key part -- the actual doc generation. I am not sure I personally can even test the last bit without accidentally pushing something. I think that, if nobody else sees an issue with it and we believe it's working, it can go into master for 3.1.

**Comment by HyukjinKwon** [comment]
There's nothing confusing to "Affected Version" on improvement type. It just targets where you want that fix to be landed. The general guide line is bug fixes only but it's judged case-by-case in practice. For example, I just backport minor documentation improvement when possible to reduce conflicts.

The only thing wrong here is I merged #27626 mistakenly with the same JIRA here and @nchammas used the same JIRA mistakenly. There was just a bit of miscommunication, which we should avoid next time.

**Comment by HeartSaVioR** [comment]
> There's nothing confusing to "Affected Version" on improvement type. It just targets where you want that fix to be landed.

In fact we don't guide like that. For example, suppose someone found a point to improve in Spark 2.4.x - they would set "Affected Version" to be 2.4.x as it may be the only version they know about, but as I see some guides on the fly we have been guiding to the latest "unreleased" minor version which they never know about.

While I don't think it represents the intention clearly on marking Affected Version on improvement, would we call it simply "the version which master branch targets to", and add it to the contribution guide? In contribution guide we only mention about bugs, no other types.

**Comment by HyukjinKwon** [comment]
This is written in the guide:

> Affects Version. For Bugs, assign at least one version that is known to exhibit the problem or need the change

Ideally they should search JIRAs and see if there are duplicates. If there are duplicate, we should resolve it as a duplicate. We can clarify more, of course. However, I don't see that this is confusing or a bad pattern yet.

**Comment by HeartSaVioR** [comment]
It only says for bugs, while we're saying improvements. Anyway I will go forward - let's not pollute the original topic.

**Comment by srowen** [comment]
FWIW, I don't think Affected Version really matters much for an improvement. You can set it to say it would be a valid improvement as of a version -- maybe it improves a thing that's only present from 2.4.0 onwards or something. But for most cases it probably 'affects' (could be applied to) lots of versions, and that isn't as interesting as in the case of a bug. Not required unless it's really important clarify, doesn't hurt in most cases either.

**Comment by srowen** [comment]
Just to be clear, any objections to merging?

**Comment by nchammas** [comment]
I don't think any committers have tried working with the proposed build instructions and/or Docker definition. Perhaps a committer could try things out to make sure it works for them?

File: docs/README.md
Comment (HyukjinKwon) [reply docs/README.md]: Hm, there's tradeoff to pin the version (test and fix it compatible first vs easily breaking). I think we haven't faced any notable problem so far to maintain those implicit compatibility across versions. What about we pin the version when actually we see some overhead to maintain this?

File: docs/README.md
Comment (HyukjinKwon) [reply docs/README.md]: I am not strongly against this change, but not very fond of it as it forces to install unnecessary dependencies. Other people might use conda to manage Python versions as an example. At least I use conda. Given that there are many ways to control this, I am not sure about documenting one way. Why don't we just mention Ruby 2 and Python 3?

Comment (nchammas) [reply docs/README.md]: We can just tell people to install Ruby 2 and Python 3, but without specifying specific versions of them we risk having contributors and committers waste time dealing with library versions that are incompatible with the version of the language/runtime they have installed, or that somehow don't work with our configuration.

It's a small risk, but I figured if we're going to use rbenv and pyenv in building the Docker image, it would make sense to recommend them for the interactive/local workflow as well. The experience is consistent, which is what's nice about using these tools, and they are "lower level" than things like Conda in that they are relatively simple shell scripts that only manage installations.

Are you OK with using rbenv and pyenv in the Docker image?

File: docs/README.md
Comment (nchammas) [reply docs/README.md]: I'm not following your concern about pinning versions. Is there some scenario or overhead from pinning versions that you are concerned about?

It seems inevitable to me that contributors will waste time debugging issues that stem from them running different versions of these libraries. Instead of waiting for those issues to arise, it seems safe and simple enough to avoid them by pinning what we use. In fact, what prompted me to open this PR is an interaction I had with you on another PR where we were seeing different behavior when trying to build the docs. So I believe we've already seen at least one example of this problem. 😄 (I linked to that interaction from this PR's description.)

Pinning our doc build dependencies is the core idea of this PR. If we don't want to do that, then I'm happy to close this PR and leave things as-is.

Comment (srowen) [reply docs/README.md]: I guess I'd default to pinning the version in the absence of a strong argument either way. I do think we've had inconsistencies in the spark-website project due to differing jekyll versions, for example.

File: docs/README.md
Comment (HyukjinKwon) [reply docs/README.md]: I think docker changes are fine.

File: docs/README.md
Comment (HyukjinKwon) [reply docs/README.md]: Okie. I am good with this change.

**Comment by srowen** [comment]
Jenkins retest this please

File: docs/README.md
Comment (HyukjinKwon) [inline docs/README.md]: Can we revert these doc changes in the documentation? Otherwise, I am good with other changes.

Comment (HyukjinKwon) [reply docs/README.md]: Or alternatively, we can have one section to describe use `rbenv` and `pyenv` separately .. that works to me too.

Comment (nchammas) [reply docs/README.md]: I moved them to their own sub-section. If we think it's distracting / not helpful, I can remove it.

File: dev/create-release/spark-rm/Dockerfile
Comment (HyukjinKwon) [inline dev/create-release/spark-rm/Dockerfile]: Okay, this part seems not tested:

```bash
Generating SQL API Markdown files.
20/03/31 06:41:42 WARN NativeCodeLoader: Unable to load native-hadoop library for your platform... using builtin-java classes where applicable
Traceback (most recent call last):
  File "/opt/spark-rm/output/spark/sql/gen-sql-api-docs.py", line 21, in <module>
    from pyspark.java_gateway import launch_gateway
  File "/opt/spark-rm/output/spark/python/lib/pyspark.zip/pyspark/__init__.py", line 51, in <module>
  File "/opt/spark-rm/output/spark/python/lib/pyspark.zip/pyspark/context.py", line 22, in <module>
ImportError: No module named threading
log4j:WARN No appenders could be found for logger (org.apache.spark.util.ShutdownHookManager).
log4j:WARN Please initialize the log4j system properly.
log4j:WARN See http://logging.apache.org/log4j/1.2/faq.html#noconfig for more info.
```

Seems the installed Python is weird. [`threading`]( https://docs.python.org/3.7/library/threading.html) is the standard Python library that has existed from Python 2 to Python 3, but seems not existent with the Python installed here.

I am going to revert this to unblock RC preparations.

**Comment by cloud-fan** [comment]
We can manually test it by changing the `ENTRYPOINT` in `Dockerfile` to `/bin/bash` and check if the python and ruby are installed correctly.

The release script doesn't work anymore after this patch because we use a non-standard way to install python/ruby (download a bash script and run it):
1. The actual work is done by user `spark-rm`, which can't access `/root/.pyenv`
2. Even if we fix the permission issue, the installed python doesn't work well with the system libraries, and have errors like `ImportError: No module named threading`

**Comment by HyukjinKwon** [comment]
Reverted at https://github.com/apache/spark/commit/4d4c3e76f6d1d5ede511c3ff4036b0c458a0a4e3.

**Comment by dongjoon-hyun** [comment]
BTW, are we using the release script in `master` branch?
We had better use the release script in `branch-3.0` to avoid `Apache Spark 2.4.2` situation.

cc @rxin since he is a release manager.

**Comment by cloud-fan** [comment]
Yea we should use the script in branch-3.0 so this doesn't block the RC. But we should still revert this PR as it doesn't work.

**Comment by dongjoon-hyun** [comment]
Sure~ Of course for reverting this PR.

**Comment by HyukjinKwon** [comment]
I think my wording made some miscommunication here:

> Let me revert this to make RC preparation easier.

I simply meant future RCs :-).

File: dev/create-release/spark-rm/Dockerfile
Comment (nchammas) [reply dev/create-release/spark-rm/Dockerfile]: Hmm, that's strange, because I was able to build enough of the Dockerfile to get past this point, and I did not hit this problem.

What I couldn't test was to run everything via `do-release-docker.sh`.

**Comment by nchammas** [comment]
> We can manually test it by changing the `ENTRYPOINT` in `Dockerfile` to `/bin/bash` and check if the python and ruby are installed correctly.
> 
> The release script doesn't work anymore after this patch because we use a non-standard way to install python/ruby (download a bash script and run it):
> 
> 1. The actual work is done by user `spark-rm`, which can't access `/root/.pyenv`
> 2. Even if we fix the permission issue, the installed python doesn't work well with the system libraries, and have errors like `ImportError: No module named threading`
> 3. It's wrong to append the `PATH`, and we should prepend it. Otherwise `python` still points to the system default which is 2.7.

This is strange since I tested this successfully via `docker build .` from within the folder containing the Dockerfile.

What I couldn't do was test building the image via `do-release-docker.sh`, which I called out in the PR description and again [in the comments](https://github.com/apache/spark/pull/27534#issuecomment-592144594).

I think the reason my test didn't see these issues is because the `do-release-docker.sh` script builds the image with some options that change the executing user from `root`.

```
run_silent "Building spark-rm image with tag $IMGTAG..." "docker-build.log" \
  docker build --no-cache -t "spark-rm:$IMGTAG" --build-arg UID=$UID "$SELF/spark-rm"
```

I guess the moral of the story is that I couldn't test `do-release-docker.sh` directly, and my substitute test of `docker build .` was not a good test since it didn't account for the `root` vs. `spark-rm` user change.

I should have more carefully tried to duplicate the `docker build` command as it appeared in `do-release-docker.sh`, and insisted more forcefully that we not merge this PR in until a committer had had a chance to try out `do-release-docker.sh` directly.

Sorry.

**Comment by nchammas** [comment]
I can try taking another stab at this change for the future, now that I have a better idea of how to test it.

However, given that this change was met with some resistance to begin with, I won't try to fix it unless there is explicit committer interest.

**Comment by cloud-fan** [comment]
This PR does have a good point to fix the dependency versions so that the script is more robust. I'm happy to see a working version of it.

**Comment by HyukjinKwon** [comment]
We can pin the version for release related ones; however, I doubt if we should do that for others e.g., CI, documentation. It's being discussed at https://github.com/apache/spark/pull/27928#pullrequestreview-385094894

**Pull request closed without merging**
